sdk/core/azure-core/inc/azure/core/internal/strings.hpp

```diff
@@ -19,24 +20,61 @@ namespace Azure { namespace Core { namespace _internal {
    */
   struct StringExtensions final
   {
+    static constexpr char ToUpper(char c) noexcept
+    {
+      return (c < 'a' || c > 'z') ? c : c - ('a' - 'A');
+    }
+
+    static constexpr char ToLower(char c) noexcept
+    {
+      return (c < 'A' || c > 'Z') ? c : c + ('a' - 'A');
+    }
+
     struct CaseInsensitiveComparator final
     {
-      bool operator()(const std::string& lhs, const std::string& rhs) const
+      bool operator()(std::string const& lhs, std::string const& rhs) const
       {
         return std::lexicographical_compare(
-            lhs.begin(), lhs.end(), rhs.begin(), rhs.end(), [](char c1, char c2) {
-              return ToLower(c1) < ToLower(c2);
+            lhs.begin(), lhs.end(), rhs.begin(), rhs.end(), [](auto l, auto r) {
+              return ToLower(l) < ToLower(r);
             });
       }
     };
 
     static bool LocaleInvariantCaseInsensitiveEqual(
-        const std::string& lhs,
-        const std::string& rhs) noexcept;
-    static std::string const ToLower(std::string const& src) noexcept;
-    static unsigned char ToLower(unsigned char const src) noexcept;
-    static std::string const ToUpper(std::string const& src) noexcept;
-    static unsigned char ToUpper(unsigned char const src) noexcept;
+        std::string const& lhs,
+        std::string const& rhs) noexcept
+    {
+      auto const rhsSize = rhs.size();
+      if (lhs.size() != rhsSize)
+      {
+        return false;
+      }
+
+      auto const lhsData = lhs.c_str();
+      auto const rhsData = rhs.c_str();
+      for (size_t i = 0; i < rhsSize; ++i)
+      {
+        if (ToLower(lhsData[i]) != ToLower(rhsData[i]))
+        {
+          return false;
+        }
+      }
+
+      return true;
+    }
+
+    static std::string ToLower(std::string src)
+    {
+      std::transform(src.begin(), src.end(), src.begin(), [](auto c) { return ToLower(c); });
+      return src;
+    }
+
+    static std::string ToUpper(std::string src)
+    {
+      std::transform(src.begin(), src.end(), src.begin(), [](auto c) { return ToUpper(c); });
+      return src;
+    }
```

Comment:
Why did we decide to move the implementation detail from a cpp file, into a header-only file?

Reply:
After static variable removal, no hard reason left to keep any implementation in a .cpp file, allowing better inlining to be used for the functions, including the functions that get called for each individual character, or from a case insensitive comparator in maps, plus I think logistically it is easier to support smaller amount of code in less files, especially if a function is short, and no .cpp file exists already.

Reply:
In general I'm a massive fan of header-only implementations of types, but Casey corrected me about implementation of header-only vs cpp - it's actually not that much better if LTCG (link time code generation) is enabled.

But as I said, I generally prefer header-only implementations when possible.

Reply:
I'd actually move some inline functions we have in headers to .cpp files - (longer ones, the ones with asciiz literals), but not these ones.